iml-wasm-components/iml-input/src/lib.rs

```diff
@@ -0,0 +1,123 @@
+#[macro_use]
+extern crate seed;
+
+extern crate combine;
+
+use combine::stream::easy::Error;
+use combine::stream::easy::Errors;
+use combine::stream::easy::Info;
+use seed::prelude::*;
+
+use iml_tooltip::TooltipPlacement;
+
+pub struct Model<T> {
+    pub value: Result<T, String>,
+    pub tooltip: iml_tooltip::Model,
+}
+
+impl<T> Default for Model<T> {
+    fn default() -> Self {
+        Self {
+            value: Err(String::new()),
+            tooltip: iml_tooltip::Model {
+                error_tooltip: true,
+                open: true,
+                placement: TooltipPlacement::Bottom,
+                ..Default::default()
+            },
+        }
+    }
+}
+
+#[derive(Clone)]
+pub enum Msg {
+    FlipTooltip,
+    InputChanged(String),
+}
+
+pub type ParserFn<T> = fn(_: &str) -> Result<T, Errors<char, &str, usize>>;
+
+fn format_err(err: &Errors<char, &str, usize>) -> String {
+    let mut un = vec![];
+    let mut ex = vec![];
+
+    fn format_info(i: &Info<char, &str>) -> String {
+        match i {
+            Info::Token(t) => {
+                if *t != '"' {
+                    format!("\"{}\"", t)
+                } else {
+                    format!("{}", t)
+                }
+            }
+            Info::Static(s) => String::from(*s),
+            Info::Owned(s) => String::from(s),
+            Info::Range(r) => String::from(*r),
+        }
+    }
+
+    for x in &err.errors {
```

Comment:
Looping like this works fine. As an alternative, you can eliminate the mutable variables `un` and `ex` above by folding:
```
let (un, ex) = &err
        .errors
        .iter()
        .fold((vec![], vec![]), |(mut un, mut ex), x| match x {
            Error::Unexpected(i) => {
                un.push(format_info(&i));
                (un, ex)
            }
            Error::Expected(i) => {
                ex.push(format_info(&i));
                (un, ex)
            }
            _ => (un, ex),
        });
```

Reply:
Yeah, but it looks less readable :)

Reply:
To be clear, i'm ok with the for loop. I shared the fold as an alternative because it reduces mutable state within the scope. Personally I find the fold a bit easier to reason about because I don't have to think about the possibility of `un` and `ex` changing within the scope of the function since they are immutable.